File: 0101-0200/0198.House_Robber.cpp
//0ma






class Solution {
public:
    int rob(vector<int>& nums) {
        int n = nums.size();

        if (n == 1) {
            return nums[0];
        }

        vector<int> dp(n, 0);

        dp[0] = nums[0];
        dp[1] = max(nums[0], nums[1]);

        for (int i = 2; i < n; i++) {
            dp[i] = max(dp[i - 1], nums[i] + dp[i - 2]);
        }

        return dp[n - 1];        
    }
};






//0ms






class Solution {
public:
    int rob(vector<int>& nums) {
        int size = nums.size();
        if (size == 0) {
            return 0;
        }
        vector<int> dp(size + 1);
        dp[0] = 0;
        dp[1] = nums[0];

        for (int i = 2; i <= size; i++) {
            dp[i] = max(nums[i - 1] + dp[i - 2], dp[i - 1]);
        }

        return dp[size];
    }
};

File: 0101-0200/0198.House_Robber.rs
//0ms




impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        nums.iter().fold((0, 0), |(a, b), x| (b, b.max(a + x))).1
    }
}



//0ms




impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        match nums.len() {
            0 => 0,
            1 => nums[0],
            2 => nums[0].max(nums[1]),
            3 => nums[0].max(nums[1]).max(nums[0] + nums[2]),
            n => {
                let mut a = nums[0];
                let mut b = nums[1];
                let mut c = nums[0] + nums[2];
                for index in 3..n {
                    (a, b, c) = (b, c, (a + nums[index]).max(b + nums[index]));
                }
                b.max(c)
            }
        }
    }
}





//0ms





impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        if nums.is_empty() {
            return 0;
        }
        if nums.len() == 1 {
            return nums[0];
        }

        let mut prev_prev = nums[0];
        let mut prev = nums[0].max(nums[1]);

        for i in 2..nums.len() {
            let current = prev_prev + nums[i];
            prev_prev = prev;
            prev = current.max(prev);
        }

        prev
    }
}




//0ms





use std::cmp::max;
impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let n = nums.len();
        if(n == 1){
            return nums[0];
        }else if(n == 2){
            return max(nums[0],nums[1]);
        }
        let (mut a,mut b,mut c) = (nums[0],max(nums[0],nums[1]),0);
        for i in (2..n){
            c = max(nums[i] + a, b);
            a = b;
            b = c;
        }
        return c;
    }
}
